// Resolution dns asynchrone

// Puprose:
// Exemple simple de r√©solution dns asynchrone
// Illustre cas ecole de r√©solution √† travers un boost::asio::io
// Utilisation d'un seul io(io_context) pour toutes les r√©solutions
// Il faut un io.restart() pour accepter de nouvelles r√©solutions
// [NEW] work_guard pour √©viter que io.run() ne sorte trop t√¥t
// [NEW] Encapsulation dans une seule classe Resolver

// Author : Toufik ABDELMOUMENE

// File: resolver.cpp


#include <boost/asio.hpp>
#include <iostream>

using boost::asio::ip::tcp;

void on_resolve(const boost::system::error_code& ec,
                tcp::resolver::results_type results)
{
    if (ec) {
        std::cerr << "Erreur de r√©solution: " << ec.message() << "\n";
        return;
    }

    std::cout << "Solve OK:\n";
    for (const auto& entry : results) {
        std::cout << entry.endpoint().address().to_string() << "\n";
    }
}

int main() {
#if defined(_WIN32)
    SetConsoleOutputCP(CP_UTF8);
#endif  
    boost::asio::io_context io;       
    tcp::resolver resolver(io);  

    // Pour emp√™cher io.run() de s‚Äôarr√™ter imm√©diatement.
     auto work_guard = boost::asio::make_work_guard(io);

 while(1)
    {
    try {

        std::string host = "www.google.com";
        std::string port = "80";
        std::cout << "üîπ Enter host" << std::endl;
        std::cin >> host;

        if(host == "000"){
            work_guard.reset(); // indique que io peut s‚Äôarr√™ter.
            break;
        }
        auto lamb = [host](const boost::system::error_code& ec,
               tcp::resolver::results_type results) {
            if (ec) {
                std::cerr << "Erreur pour " << host << ": " << ec.message() << "\n";
                return;
            }

            std::cout << "‚úÖ Solving " << host << " OK:\n";
            for (const auto& entry : results) {
                std::cout << entry.endpoint().address().to_string() << "\n";
            }
        };
        // D√©marre la r√©solution asynchrone
        resolver.async_resolve(host, port, lamb);
        

            // üîß Red√©marrer le contexte pour accepter de nouvelles op√©rations
            io.restart();

            // üîÅ Relancer le traitement des √©v√©nements asynchrones
            io.run();
    }
    catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << "\n";
    }
    }
    return 0;
}
